Adding 'namespace' for prometheus metrics

alexbarcelo opened the pull request:
## Description:

Add a "namespace" for all the component metrics. This is a breaking change, motivated by "official best practices" [see Prometheus documentation](https://prometheus.io/docs/practices/naming/)

> A metric name [...] should have a (single-word) application prefix relevant to the domain the metric belongs to. The prefix is sometimes referred to as namespace by client libraries. For metrics specific to an application, the prefix is usually the application name itself.

The code could be tweaked to be non-breaking. IMHO, it is more sane and well-organized to force a top-level namespace and share that by all the metrics --instead of starting the metric name by the hass domain-- but it is arguably the best approach. Before adding the documentation I wanted to discuss that and listen to other opinions. If there is consensus in a different solution, I have no problem modifying this PR to follow the guidelines.

## Example entry for `configuration.yaml` (if applicable):
```yaml
prometheus:
  namespace: "hass"  # this happens to be the default namespace name
```

## Result:

Going to http://hass.example.net:8123/api/prometheus shows:

    ...
    # HELP hass_device_tracker_state State of the device tracker (0/1)
    # TYPE hass_device_tracker_state gauge
    hass_device_tracker_state{entity="device_tracker.04xxxxxxxx9d",friendly_name="04xxxxxxxx9d"} 1.0
    hass_device_tracker_state{entity="device_tracker.androidffxxxxxxxxxxxx2f",friendly_name="android-ffxxxxxxxxxxxxxx2f"} 1.0
    ...

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io): home-assistant/home-assistant.github.io#5152

pvizeli commented on homeassistant/components/prometheus.py:
```diff
@@ -29,8 +30,13 @@
+
 CONFIG_SCHEMA = vol.Schema({
-    DOMAIN: recorder.FILTER_SCHEMA,
+    DOMAIN: recorder.FILTER_SCHEMA.extend({
```
Use the origin filter schema from https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/helpers/entityfilter.py

alexbarcelo commented:
I messed up a little bit at the middle (I was not in my main development machine, and did some silly mistakes).

While looking into other components and seeing the entity_filter helper, I felt that moving the filter into its own entry in the settings is a better approach. That means more breaking changes, although the code is easier to mainteinance. The settings will be like this:

    prometheus:
      filter:
        <filter settings here>
      namespace: "name"

Although the default

    prometheus:

will keep working because everything else is optional.

alexbarcelo commented:
@pvizeli I think that the changes done were the ones you were requesting. Could you confirm that it looks good? At least that specific aspect.

pvizeli commented:
No my change request are the same: https://github.com/home-assistant/home-assistant/blob/3fd61d8f456f14796e41dc05c77a7ab60861eb84/homeassistant/components/alexa/__init__.py#L34

You should not access to a other component from your componenent. That break the interface logic and as next with the helper, we can provide a config feature they work over multible component in same way. We need migrate all the old ways to this helper.

alexbarcelo commented:
I don't understand... Are you refering to an alexa component? I only touched Prometheus stuff... in fact, your change request was related to a stuff that was already there (but given that I was changing that part, it was the moment to fix it). I think that there is no reference to other component, or I am missing something.

alexbarcelo commented on homeassistant/components/prometheus.py:
```diff
@@ -31,13 +28,15 @@
 CONFIG_SCHEMA = vol.Schema({
-    DOMAIN: FILTER_SCHEMA.extend({
+    DOMAIN: vol.All({
+        vol.Optional(CONF_FILTER, default={}): entityfilter.FILTER_SCHEMA,
```
@pvizeli This is what you asked, isn't it? Using the `entityfilter.FILTER_SCHEMA` at this point (instead of the previous other-component-stuff-that-may-break).

fabaff commented:
Could you please update the dependency (`REQUIREMENTS`)? You have a working setup and this would ensure that the upgrade works. Thnaks.

alexbarcelo commented:
@fabaff Updated to `prometheus_client=0.2.0` and restarted my local home assistant installation to test that everything was ok.

Main functionality is working as expected --let me remind that there are breaking changes on the naming of metrics, albeit I defend that there is "a greater good" behind the rename.

I don't know why coveralls is complaining now (or at all). Maybe it has to do with my f*** up when I started on the wrong branch? (shame on me).

balloob commented:
You can ignore coveralls. It has already been fixed upstream.

You removed the PR instructions that you need to follow to get Travis to pass… run `script/gen_requirements_all.py`

alexbarcelo commented on homeassistant/components/prometheus.py:
```diff
@@ -29,8 +28,15 @@
-    DOMAIN: recorder.FILTER_SCHEMA,
+    DOMAIN: vol.All({
+        vol.Optional(CONF_FILTER, default={}): entityfilter.FILTER_SCHEMA,
+        vol.Optional(CONF_PROM_NAMESPACE, default=PROM_NAMESPACE): cv.string,
```
you mean that the metrics name, by default, should not have _application prefix_ at all? The reason for that is to have the "previous" (current, actual) behaviour of prometheus component by default?

alexbarcelo committed:
Updating prometheus client version

alexbarcelo committed:
Using `entity_filter` as filter mechanism

alexbarcelo committed:
New optional `namespace` configuration

alexbarcelo commented:
I redid the branch because there were so many mistakes on my part that commit history was a mess.

@pvizeli I think that I addressed your issues at a certain point but that was not clear. On the new commits, if 72351dcd176cc835248b76b516114bf21f9fc4d0 is not what you were asking, then please tell me so because that means that I still not know what to change :( That commit is in fact unrelated to the purpose of this PR, but it was a good moment to improve it.

@balloob I assumed that you wanted to have no default and, if no `namespace` is given, maintain the current behaviour. That is what is done (just tested it): the "metrics prefix" will be `<namespace>_` or empty, depending if `namespace` settings is set.

The breaking changes are related to the use of `filter` settings. Again, not my main purpose of this PR, but collateral ~~damage~~improvements.

alexbarcelo commented:
Is this PR good to go? Or it still has issues?
